### PR TITLE
console: allow configuring statement timeout on console RawSQL page (close #4998)

### LINKED ISSUE #4998: add a configurable statement_timeout on console raw sql page
To avoid accidently running a very heavy query on the database, the console can set a configurable transaction level `statement_timout` of 10 seconds.

Screenshot from 2020-06-08 17-01-06

If a user updates this value, the changed value should be persisted in local storage and used when the raw sql page is accessed subsequently.

The tooltip for the statement timeout input should be: `Abort queries that take longer than the specified time`

A statement_timout for a transaction can be set by adding the following SQL statement before the actual query (the server runs each query in a transaction so nothing needs to be done for that)

```sql
SET LOCAL statement_timeout = 10000;   # note that statement_timout expects input in milliseconds
```

### PR DESCRIPTION
**Pull request by gokulchandra**
### Description

Allows users to configure statement level timeouts. 
More info here #4998 

### Changelog

- [x] `CHANGELOG.md` is updated with user-facing content relevant to this PR. 

### Affected components

- [ ] Server
- [x] Console
- [ ] Docs
- [ ] Community Content
- [ ] Tests

### Related Issues

#4998

### CODE REVIEW

**Comment by rikinsk** [comment]
I resolved conflicts with master and took the liberty to make a couple of minot UI changes. 

I also set the default initial value of the timeout as 10 seconds

**Comment by rikinsk** [comment]
@beerose Just realised we might have to think about how to handle this if `This is a migration` flag is checked. We wouldnt want the statement timeout to be added to the migrations file. Maybe we dont add the statement timeout when the SQL statement is marked as a migration?

**Comment by beerose** [comment]
> Maybe we dont add the statement timeout when the SQL statement is marked as a migration?

That sounds ok. If we do this, we probably could make timeout input disabled with the tooltip saying that it's not possible to set a timeout for migrations?

Another solution that comes to my mind is to take advantage of `skip_execution` flag and generate migration in a separate API call.

**Comment by rikinsk** [comment]
>  we probably could make timeout input disabled with the tooltip saying that it's not possible to set a timeout for migrations?
This sounds good enough

**Comment by beerose** [comment]
@gokulchandra Would you mind adding functionality mention in the previous comments  (https://github.com/hasura/graphql-engine/pull/5045#issuecomment-645964449, https://github.com/hasura/graphql-engine/pull/5045#issuecomment-645968494)?

What would need to be done is:
* In `executeSQL` function — don't add timeout SQL if `isMigration` (first argument) is true.
* In `RawSQL` component — based on `isMigrationChecked` prop enable or disable setting timeout statement. In case of disabling this section, [here](https://github.com/hasura/graphql-engine/pull/5045/files#diff-bbe6886a78197981286f8e23813cb0ceR432) we could have class `.disabled` (from `Common.scss`) and disable the input. We can also have a tooltip saying that it's not possible to set a timeout when migration is checked. 

It would be also good to extract `getStatementTimeoutSection` to a separate component in a new file.

**Comment by gokulchandra** [comment]
@beerose Sure.. Thanks for detailing that.. I will update the PR accordingly

**Comment by gokulchandra** [comment]
> @gokulchandra Would you mind adding functionality mentioned in the previous comments ([#5045 (comment)](https://github.com/hasura/graphql-engine/pull/5045#issuecomment-645964449), [#5045 (comment)](https://github.com/hasura/graphql-engine/pull/5045#issuecomment-645968494))?
> 
> What would need to be done is:
> 
>     * In `executeSQL` function — don't add timeout SQL if `isMigration` (first argument) is true.
> 
>     * In `RawSQL` component — based on `isMigrationChecked` prop enable or disable setting timeout statement. In case of disabling this section, [here](https://github.com/hasura/graphql-engine/pull/5045/files#diff-bbe6886a78197981286f8e23813cb0ceR432) we could have class `.disabled` (from `Common.scss`) and disable the input. We can also have a tooltip saying that it's not possible to set a timeout when migration is checked.
> 
> 
> It would be also good to extract `getStatementTimeoutSection` to a separate component in a new file.

@beerose and @rikinsk updated the PR to include the above functionality..

File: console/src/components/Services/Data/RawSQL/Actions.js [hunk 1]
```diff
@@ -160,6 +176,16 @@ const executeSQL = (isMigration, migrationName) => (dispatch, getState) => {
   );
 };
 
+export const setRawSqlTimeout = timeoutInSeconds => {
```
[the remark this answers is not in the document]
Comment (gokulchandra) [reply console/src/components/Services/Data/RawSQL/Actions.js:179]: Out of curiosity: Is there a particular reason why we'd prefer having `localStorage` logic within the component? 
I was under the assumption that all side effects(API calls, IO) are handled in a single place within the actions..

[the remark this answers is not in the document]
Comment (beerose) [reply console/src/components/Services/Data/RawSQL/Actions.js]: So I was going to leave another comment — we should avoid using Redux as a local state cache. `timeout` is only used by single page — `RawSQL` so we can as well have it as a state variable in `RawSQL.js` and then use it in `submitSQL`, right?

[the remark this answers is not in the document]
Comment (gokulchandra) [reply console/src/components/Services/Data/RawSQL/Actions.js]: Fair enough.. I've moved that state to the component.

File: CHANGELOG.md [hunk 2]
```diff
@@ -22,11 +22,13 @@
 
 - server: fix introspection when multiple actions defined with Postgres scalar types (fix #5166) (#5173)
 - console: allow manual edit of column types and handle array data types (close #2544, #3335, #2583) (#4546)
+- console: allow configurable statement_timeout on console raw sql page (close #4998) (#5045)
```
Comment (beerose) [inline CHANGELOG.md:25]:
```suggestion
- console: allow configuring statement timeout on console RawSQL page (close #4998) (#5045)
```

Allow configuring statement timeout on console RawSQL page.

File: CHANGELOG.md [hunk 3]
```diff
@@ -163,6 +164,7 @@ hasura seed apply --file 1234_add_some_seed_data.sql
 - console: fix visiting view modify page overwriting raw sql content (fix #4798) (#4810)
 - console: add help button and move about page to settings (#4848)
 - console: add new sidebar icon that separates enums from tables (fix #4984) (#4992)
+- console: fix "Cannot read property 'foldable'" runtime error in `Browse Rows` page (fix #4907) (#5016)
```
Comment (beerose) [inline CHANGELOG.md:167]:
```suggestion
```

**Comment by marionschleifer** [comment]
@gokulchandra thank you for contributing to Hasura 🙏 we'd love to send you some Hasura swag as a thank you. If you're interested, please send an email to marion@hasura.io and we'll take it from there 🙂

**Pull request merged**